crates/ra_hir/src/ty/tests.rs

```diff
@@ -287,6 +287,29 @@ fn test(x: &str, y: isize) {
     );
 }
 
+#[test]
+fn infer_array() {
+    check_inference(
+        r#"
+fn test(x: &str, y: isize) {
+    let a = [x];
+    let b = [a, a];
+    let c = [b, b];
+
+    let d = [y];
+    let e = [d, d];
+    let f = [e, e];
+
```

Comment:
here's an expression that checks whether the element types are actually unified:
```
[y, 1, 2, 3]
```
the 1, 2 and 3 should be inferred to `isize`, and I think with your current code they won't.

Reply:
you could also add the case where the array is empty:
```
let x: [u8; 0] = [];
```
we should still infer the element type on the right side. I think this should already work. This would also test array type references, which you implemented but didn't use in this test (the `[u8; 0]` part).